persistent-mysql/Database/Persist/MySQL.hs

```diff
@@ -1034,10 +1037,25 @@ insertOnDuplicateKeyUpdate
 insertOnDuplicateKeyUpdate record =
   insertManyOnDuplicateKeyUpdate [record] []
 
--- | This wraps values of an Entity's 'EntityField', making them have the same
--- type. This allows them to be put in lists.
+-- | This type is used to determine how to update rows using MySQL's
+-- @INSERT ON DUPLICATE KEY UPDATE@ functionality, exposed via
+-- 'insertManyOnDuplicateKeyUpdate' in the library.
 data SomeField record where
   SomeField :: EntityField record typ -> SomeField record
+  -- ^ Copy the field directly from the record.
+  CopyUnlessEq :: PersistField typ => EntityField record typ -> typ -> SomeField record
+  -- ^ Only copy the field if it is not equal to the provided value.
+
+-- | Copy the field into the database only if the value in the
+-- corresponding record is non-@NULL@.
+copyUnlessNull :: PersistField typ => EntityField record (Maybe typ) -> SomeField record
+copyUnlessNull field = CopyUnlessEq field Nothing
+
+-- | Copy the field into the database only if the value in the
+-- corresponding record is non-empty, where "empty" means the Monoid
+-- definition for 'mempty'. Useful for 'Text', 'String', 'ByteString', etc.
+copyUnlessEmpty :: (Monoid typ, PersistField typ) => EntityField record typ -> SomeField record
+copyUnlessEmpty field = CopyUnlessEq field mempty
 
```

Comment:
Both of the introduced functions are only seem to be applicable for `insertManyOnDuplicateKeyUpdate`. Can you mention that in the haddock comments? Also can you add the example you mentioned here: https://github.com/yesodweb/persistent/pull/693#issuecomment-326688351 and explain it's usecase in  a similar way we have done it here: [persistent doc](https://www.stackage.org/haddock/lts-7.0/persistent-2.6/Database-Persist.html#v:-43--61-..)